### PR TITLE
Handle `last_insert_id()` and `database()` in vtgate

### LINKED ISSUE #3668: Improve treatment of LAST_INSERT_ID() in Vitess
For a table defined as
```
CREATE TABLE `my_table` (
  `id` int(11) NOT NULL AUTO_INCREMENT,
  `col2` bigint(20) unsigned DEFAULT NULL,
  `col3` bigint(20) unsigned DEFAULT NULL,
  `type` enum(…,'enum_value',…) NOT NULL DEFAULT '...',
  PRIMARY KEY (`id`)
) ENGINE=InnoDB AUTO_INCREMENT=9999 DEFAULT CHARSET=latin1 ROW_FORMAT=DYNAMIC
```

I connect to a keyspace explicitly and see the following:
```
vt_app@localhost keyspace@master> insert into my_table (col2,col3,type) values (10000000000,10000000000,'enum_value');
Query OK, 1 row affected (0.01 sec)
vt_app@localhost keyspace@master> select last_insert_id();
+------------------+
| last_insert_id() |
+------------------+
|                0 |
+------------------+
(id=2353 is the row inserted)
```

I expect to get back from `LAST_INSERT_ID()` the value 2353 but don't get it.

Explanation is:
* using autocommit
* this means that the insert connection for statment 1 and the select statement for statement 2 are not necessarily using the same mysql connection.

Current workaround:
* wrap in `BEGIN; .....; COMMIT;` to use the same connection

Improvement in behaviour (suggested by @sougou) is to record the value in the client session
and in subsequent statements replace `LAST_INSERT_ID()` by the value that was previously recorded.

This issue to to record this problem and a mention a possible solution suggested by @sougou.

### LINKED ISSUE #5417: last_insert_id is broken
When filing a bug, please include the following headings if
possible. Any example text in this template can be deleted.

#### Overview of the Issue

The following query was recently made illegal in Vitess (since the merge of #5268)

```SELECT last_insert_id()``` 

#### Reproduction Steps

Steps to reproduce this issue, example:

1. Deploy the following `vschema`:

```
{
	"sharded": false,
	"tables": {
		"messages": {},
		"tokens": {}
	}
}
```

1. Deploy the following `schema`:

```
CREATE TABLE t (id INTEGER AUTO_INCREMENT PRIMARY KEY, f VARCHAR(1));
```

2. Insert data using this query:
```
INSERT INTO t(f) VALUES('a');
```

3. Try to view the inserted auto incremented id:
```
SELECT last_insert_id()
```

you'll see the error: `unsupported: LAST_INSERT_ID is only allowed for unsharded keyspaces`

I believe this is not the correct behaviour. I did a little writeup of how LII() is supposed to work here: https://gist.github.com/systay/09240a11ec12d3c63c48b39e1b47969b

### PR DESCRIPTION
**Pull request by systay**
Fixes #3668 
Fixes #5417

### CODE REVIEW

File: go/vt/vtgate/planbuilder/expression_rewriting.go [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+		return true, nil
+	}, in)
+
+	if err != nil {
+		return nil, false, err
+	}
+
+	for _, node := range matchingNodes {
+		in = sqlparser.ReplaceExpr(in, node, bindVarExpression())
+	}
```
[the remark this answers is not in the document]
Comment (systay) [reply go/vt/vtgate/planbuilder/expression_rewriting.go]: I added subquery tests, and they seem to work

File: go/vt/vtgate/planbuilder/expression_rewriting.go [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+import (
+	"vitess.io/vitess/go/vt/proto/vtrpc"
+	"vitess.io/vitess/go/vt/sqlparser"
+	"vitess.io/vitess/go/vt/vterrors"
+	"vitess.io/vitess/go/vt/vtgate/engine"
+)
+
+// RewriteResult contains the rewritten expression and meta information about it
+type RewriteResult struct {
```
[the remark this answers is not in the document]
Comment (systay) [reply go/vt/vtgate/planbuilder/expression_rewriting.go]: I like having this intermediate struct to make it easy to write unit tests for the expression rewriting without having to first create a `primitiveBuilder`

File: go/vt/vtgate/planbuilder/builder.go [hunk 3]
```diff
@@ -267,20 +267,20 @@ func Build(query string, vschema ContextVSchema) (*engine.Plan, error) {
...
+		instruction, needsLastInsertID, needsDBName, err = buildSelectPlan(stmt, vschema)
 	case *sqlparser.Insert:
-		plan.Instructions, err = buildInsertPlan(stmt, vschema)
+		instruction, needsLastInsertID, needsDBName, err = buildInsertPlan(stmt, vschema)
 	case *sqlparser.Update:
-		plan.Instructions, err = buildUpdatePlan(stmt, vschema)
+		instruction, needsLastInsertID, needsDBName, err = buildUpdatePlan(stmt, vschema)
 	case *sqlparser.Delete:
-		plan.Instructions, err = buildDeletePlan(stmt, vschema)
+		instruction, err = buildDeletePlan(stmt, vschema)
```
Comment (sougou) [inline go/vt/vtgate/planbuilder/builder.go]: i think delete could also need last insert id.

File: go/vt/vtgate/planbuilder/expression_rewriting.go [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+	if err != nil {
+		return nil, err
+	}
+	out.UpdateBindVarNeeds(pb)
+	return out.Expression, nil
+}
+
+// Rewrite will rewrite an expression. Currently it does the following rewrites:
+//  - `last_insert_id()` => `:vtlastid`
+//  - `database()`       => `:vtdbname`
```
Comment (sougou) [inline go/vt/vtgate/planbuilder/expression_rewriting.go]: Need to update these to use the new bind var names.

**Pull request merged**